src/hotspot/cpu/riscv/nativeInst_riscv.cpp

```diff
@@ -39,83 +39,612 @@
+  bool has_address_stub();
+
+  static void set_stub_address_destination_at(address dest, address value);
+  static address stub_address_destination_at(address src);
+ public:
+
+  static NativeFarCall* at(address addr);
+  static bool is_at(address addr);
+  static bool is_call_before(address return_address);
+  static void insert(address code_pos, address entry);
+  static void replace_mt_safe(address instr_addr, address code_buffer);
+};
+
+address NativeFarCall::destination() const {
+  address addr = addr_at(0);
+  assert(NativeFarCall::is_at(addr), "unexpected code at call site");
+
+  address destination = MacroAssembler::target_addr_for_insn(addr);
+
+  CodeBlob* cb = CodeCache::find_blob(addr);
+  assert(cb && cb->is_nmethod(), "sanity");
+  nmethod *nm = (nmethod *)cb;
+  assert(nm != nullptr, "Sanity");
+  assert(nm->stub_contains(destination), "Sanity");
+  assert(destination != nullptr, "Sanity");
+  return stub_address_destination_at(destination);
+}
+
+address NativeFarCall::reloc_destination(address orig_address) {
+  address call_addr = addr_at(0);
+
+  CodeBlob *code = CodeCache::find_blob(call_addr);
+  assert(code != nullptr, "Could not find the containing code blob");
+  address stub_addr = trampoline_stub_Relocation::get_trampoline_for(call_addr, (nmethod*)code);
+
+  if (stub_addr != nullptr) {
+    stub_addr = MacroAssembler::target_addr_for_insn(call_addr);
+  }
+  return stub_addr;
+}
+
+void NativeFarCall::set_destination(address dest) {
+  address addr = addr_at(0);
+  assert(NativeFarCall::is_at(addr), "unexpected code at call site");
+  Unimplemented();
+}
+
+void NativeFarCall::verify() {
+  assert(NativeFarCall::is_at(addr_at(0)), "unexpected code at call site");
+}
+
+void NativeFarCall::print() {
+  assert(NativeFarCall::is_at(addr_at(0)), "unexpected code at call site");
+  tty->print_cr(PTR_FORMAT ": auipc,ld,jalr x1, offset/reg, ", p2i(addr_at(0)));
+}
+
+bool NativeFarCall::set_destination_mt_safe(address dest, bool assert_lock) {
+  assert(NativeFarCall::is_at(addr_at(0)), "unexpected code at call site");
+  assert(!assert_lock ||
+         (Patching_lock->is_locked() || SafepointSynchronize::is_at_safepoint()) ||
+         CompiledICLocker::is_safe(addr_at(0)),
+         "concurrent code patching");
+
+  address call_addr = addr_at(0);
+  assert(NativeFarCall::is_at(call_addr), "unexpected code at call site");
+
+  address stub_addr = stub_address();
+
+  if (stub_addr != nullptr) {
+    set_stub_address_destination_at(stub_addr, dest);
+    return true;
+  }
+
+  return false;
+}
+
+bool NativeFarCall::reloc_set_destination(address dest) {
+  address call_addr = addr_at(0);
+  assert(NativeFarCall::is_at(call_addr), "unexpected code at call site");
+
+  CodeBlob *code = CodeCache::find_blob(call_addr);
+  assert(code != nullptr, "Could not find the containing code blob");
+  address stub_addr = trampoline_stub_Relocation::get_trampoline_for(call_addr, (nmethod*)code);
+
+  if (stub_addr != nullptr) {
```

Comment:
Yes, it can, no.
Because we never want to execute anything more in `Relocation::pd_set_call_destination()` during NativeFarCall.
NativeShortCall (original) do a lot of weird things which seems a bit wrong/unnecessary.
I did not want to sort that out as I'm proposing not using it.